col/src/main/java/vct/col/ast/util/NameScanner.scala

```diff
@@ -60,15 +86,39 @@ class NameScanner extends RecursiveVisitor[AnyRef](null, null) {
     }
   }
 
+  /** If the name is in `additionalNames`, or there is a regular declaration with the same name, the types must match
+    * If there is a free variable with the same name, the types must match
+    * If there is no decl nor a free variable, we found a new free variable, which is added to the `freeNames` list
+    *
+    * @param additionalNames can be used to pass in given or yields scopes.
+    * @return true if `name` is the name of a free variable.
+    */
+  private def checkName(name: String, typ: Type, additionalNames: Function1[String, Option[DeclarationStatement]]): Boolean =
```

Comment:
`Function1` can just be `String => Option[DeclarationStatement]` I think? Or is there a subtle difference?

Reply:
And shouldn't it be a map? I guess that's applicable also haha

Reply:
I didn't know how to write a function type, `string => option` is what I intended. I guess ideally it would be a map, but since we need to take nesting into account stacks of maps are used. Therefore a function is better here for flexibility. (Or does scala understand that Map[String, Option[Decl]] can also be a function? When I tried it it just gave errors at the call sites)